timesketch/api/v1/resources/archive.py

```diff
@@ -462,58 +478,144 @@ def _unarchive_sketch(self, sketch: Sketch):
         return HTTP_STATUS_CODE_OK
 
     def _archive_sketch(self, sketch: Sketch):
-        """Unarchives a sketch by opening up all indices and removing labels.
+        """Archives a sketch. This involves:
+        1. Setting the sketch status to 'archived'.
+        2. Setting the status of its 'ready' timelines to 'archived'.
+        3. For each SearchIndex associated with these timelines:
+           If all timelines using that SearchIndex are now 'archived',
+           attempt to close the index in OpenSearch.
+           If the OpenSearch index is successfully closed (or confirmed not found),
+           set the SearchIndex's status in the database to 'archived'.
+           Otherwise, the SearchIndex status is left unchanged (e.g., 'ready'),
+           allowing tools like tsctl to identify potential issues.
 
         Args:
-            sketch: Instance of timesketch.models.sketch.Sketch
+            sketch (Sketch): Instance of timesketch.models.sketch.Sketch
         """
         if sketch.get_status.status == "archived":
             abort(
                 HTTP_STATUS_CODE_BAD_REQUEST,
-                "Unable to archive a sketch that was already archived "
-                "(sketch: {:d})".format(sketch.id),
+                f"Sketch {sketch.id} is already archived.",
             )
 
+        # Check if any timeline is in a non-archivable state
+        non_archivable_timeline_statuses = ["processing", "new", "fail", "timeout"]
+        for timeline in sketch.timelines:
+            timeline_status = timeline.get_status.status
+            if timeline_status in non_archivable_timeline_statuses:
+                abort(
+                    HTTP_STATUS_CODE_BAD_REQUEST,
+                    f"Cannot archive sketch {sketch.id}. Timeline '{timeline.name}' "
+                    f"(ID: {timeline.id}) is in '{timeline_status}' state, which "
+                    f"prevents archival. Please ensure all timelines are 'ready' or "
+                    f"'archived'.",
+                )
+
         labels_to_prevent_deletion = current_app.config.get(
-            "LABELS_TO_PREVENT_DELETION", []
+            "LABELS_TO_PREVENT_DELETION",
+            [],  # Assuming same config key for archival prevention
         )
 
         for label in labels_to_prevent_deletion:
             if sketch.has_label(label):
                 abort(
                     HTTP_STATUS_CODE_FORBIDDEN,
-                    "A sketch with the label {:s} cannot be " "archived.".format(label),
+                    f"A sketch with the label {label} cannot be archived.",
                 )
 
         sketch.set_status(status="archived")
+        logger.info("Sketch %s status set to 'archived'.", sketch.id)
```

Comment:
I know this was not part of your PR, but when we already do such big changes: Shouldn't this be the last step after all sub-components are taken care of?